File: 20. Valid Parentheses/Solution.java
class Solution {
    public boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        Map<Character, Character> b = new HashMap<>();
        b.put('(', ')');
        b.put('[', ']');
        b.put('{', '}');
        char[] ch = s.toCharArray();
        for (char c : ch) {
            if (b.containsKey(c)) {
                stack.offerLast(c);
            } else {
                if (stack.isEmpty()) return false;
                if (b.get(stack.peekLast()) != c) return false;
                stack.pollLast();
            }
        }
        return stack.isEmpty();
    }
}

File: 20. Valid Parentheses/Solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack, b = [], dict([['(', ')'], ['[', ']'], ['{', '}']])
        for c in s:
            if c in b:
                stack.append(c)
            else:
                if not stack: return False
                if b[stack[-1]] != c: return False
                stack.pop()
        return not stack